zkvm/src/predicate.rs

```diff
@@ -181,75 +156,221 @@ impl Into<CompressedRistretto> for Predicate {
+            prog.encode(&mut buf);
+            t.commit_bytes(b"prog", &buf);
+        }
 
-        // dummy predicates
-        let preds = vec![
-            Predicate::Opaque(gens.B.compress()),
-            Predicate::Opaque(gens.B_blinding.compress()),
-        ];
+        let mut leaves = Vec::new();
+        for prog in progs.iter() {
+            let blinding = rand::thread_rng().gen::<[u8; 32]>();
+            let blinding_leaf = PredicateLeaf::Blinding(blinding);
+            let program_leaf = PredicateLeaf::Program(ProgramItem::Program(prog.clone()));
+            if blinding[0] & 1 == 0 {
+                leaves.push(blinding_leaf);
+                leaves.push(program_leaf);
+            } else {
+                leaves.push(program_leaf);
+                leaves.push(blinding_leaf);
+            }
+        }
+        leaves
+    }
 
-        let pred = Predicate::disjunction(preds.clone()).unwrap();
-        let op = pred.prove_disjunction(&preds);
-        assert!(op.verify().is_ok());
+    /// Get the predicate leaves.
+    pub fn get_leaves(&self) -> &Vec<PredicateLeaf> {
+        &self.leaves
     }
 
-    #[test]
-    fn valid_disjunction2() {
-        let gens = PedersenGens::default();
+    /// Get the predicate leaf at an index.
+    pub fn get_leaf(&self, index: usize) -> &PredicateLeaf {
+        &self.leaves[index]
+    }
+
+    /// Get the verification key.
+    pub fn get_key(&self) -> &VerificationKey {
+        &self.key
+    }
+}
+
+impl CallProof {
+    pub fn serialized_length(&self) -> usize {
+        // VerificationKey is a 32-byte array
+        // MerkleNeighbor is a 32-byte array
+        32 + 4 + self.neighbors.len() * 32
+    }
+
+    pub fn decode<'a>(reader: &mut SliceReader<'a>) -> Result<Self, VMError> {
+        let mut call_proof = CallProof::default();
+        let point = reader.read_point()?;
+        let ovk = VerificationKey::from_compressed(point);
+        match ovk {
+            Some(k) => call_proof.verification_key = k,
+            None => return Err(VMError::InvalidPoint),
+        }
+        let positions = reader.read_u32()?;
+        if positions == 0 {
+            return Err(VMError::FormatError);
+        }
+        let mut neighbors = vec![];
+        let num_neighbors = 31 - positions.leading_zeros();
+        for i in 0..num_neighbors {
+            let neighbor_bytes = reader.read_u8x32()?;
+            if positions & (1 << i) == 0 {
+                neighbors.push(MerkleNeighbor::Left(neighbor_bytes));
+            } else {
+                neighbors.push(MerkleNeighbor::Right(neighbor_bytes));
+            }
+        }
+        call_proof.neighbors = neighbors;
+        Ok(call_proof)
+    }
+
+    /// Serializes the call proof to a byte array
+    pub fn encode(&self, buf: &mut Vec<u8>) {
+        encoding::write_point(self.verification_key.as_compressed(), buf);
+
+        let num_neighbors = self.neighbors.len();
+        let mut positions: u32 = 1 << num_neighbors;
```

Comment:
can we copy my description of the bit pattern and the table here as a big descriptive comment?